File: web/src/mocks/fixtures/sets.ts
// The run-set browser's fixture (`/api/v1/sets*`): aggregates over the same
// generated runs the rest of the mock serves, plus the mutable restriction and
// grant rows the access panel writes.
//
// Mirrors `storage/sets.rs` where it matters:
//   * aggregates are SUMs over every VISIBLE run of the set, never of one run;
//   * a run's pass rate is `pass_count / case_count` (the server's PASS_RATE),
//     not `pass / (pass + fail + error)`;
//   * sparklines are the last 20 runs, oldest first;
//   * `restricted` is COVERING on the browse views (a project row locks its
//     suites) and EXACT on the access payload, whose toggle owns that one row;
//   * a project or suite whose runs are all invisible DISAPPEARS rather than
//     listing with a zero count — which is what makes the 404 on the detail
//     routes indistinguishable from "never existed".

import type {
  GrantLevel,
  ProjectSetDetailResponse,
  ProjectSetView,
  SetAccessResponse,
  SetGrantView,
  SetsResponse,
  SuiteSetDetailResponse,
  SuiteSetView,
} from "@/api";
import { parseTimestamp } from "@/lib/format";
import { allRunSummaries } from "./runStats";
import { BASELINE_BY_SUITE } from "./runMeta";

/** How many recent runs a suite's sparkline covers (`SPARKLINE_RUNS`). */
const SPARKLINE_RUNS = 20;
/** How many per-suite rates a project row carries (`PROJECT_SPARK_CAP`). */
const PROJECT_SPARK_CAP = 20;

/**
 * The caller's access class — the mock's `RunVisibility`. Admins (including the
 * implicit static admin the specs browse as) are `full`; a signed-in non-admin
 * rides its own grants; anonymous callers see unrestricted sets only.
 */
export type SetViewer =
  | { kind: "full" }
  | { kind: "user"; id: string }
  | { kind: "public" };

interface RestrictionRow {
  project: string;
  /** `null` covers every suite in the project. */
  suite: string | null;
}

interface GrantRow {
  project: string;
  suite: string | null;
  user_id: string;
  username: string;
  level: GrantLevel;
  created_at: number;
  created_by: string | null;
}

/** Fixed so seeded rows read the same on every reload (see authState). */
const SEED_TIME = Date.UTC(2026, 5, 2, 9, 30, 0);

/**
 * The seeded policy.
 *
 * `support-bot` is locked at the project level and reachable by the mock's
 * `member` account, which holds `manage` over it — the pair the access panel's
 * non-admin path is exercised against. `search-rerank/ndcg-eval` is locked at
 * the SUITE level inside an otherwise open project, which is the case that
 * separates the covering answer from the exact one.
 */
function seededRestrictions(): RestrictionRow[] {
  return [
    { project: "support-bot", suite: null },
    { project: "search-rerank", suite: "ndcg-eval" },
  ];
}

function seededGrants(): GrantRow[] {
  return [
    {
      project: "support-bot",
      suite: null,
      user_id: "u_member",
      username: "member",
      level: "manage",
      created_at: SEED_TIME,
      created_by: "admin",
    },
    {
      project: "support-bot",
      suite: null,
      user_id: "u_sso",
      username: "sso.only",
      level: "view",
      created_at: SEED_TIME + 60_000,
      created_by: "admin",
    },
    {
      project: "search-rerank",
      suite: "ndcg-eval",
      user_id: "u_member",
      username: "member",
      level: "view",
      created_at: SEED_TIME + 120_000,
      created_by: "admin",
    },
  ];
}

let restrictions = seededRestrictions();
let grants = seededGrants();

/** Test hook: restore the seeded restrictions/grants. */
export function resetSets(): void {
  restrictions = seededRestrictions();
  grants = seededGrants();
}

// --- policy ----------------------------------------------------------------

/** Whether THIS exact row is restricted — what the access panel's toggle owns. */
export function setRestrictedExactly(
  project: string,
  suite: string | null,
): boolean {
  return restrictions.some((r) => r.project === project && r.suite === suite);
}

/** Whether a restriction COVERS the set: its own row, or its project's. */
function coveringRestricted(project: string, suite: string | null): boolean {
  return restrictions.some(
    (r) => r.project === project && (r.suite === null || r.suite === suite),
  );
}

const LEVEL_RANK: Record<GrantLevel, number> = { view: 1, upload: 2, manage: 3 };

/** The strongest grant this user holds over the set, project rows included. */
function coveringLevel(
  userId: string,
  project: string,
  suite: string | null,
): GrantLevel | null {
  let best: GrantLevel | null = null;
  for (const g of grants) {
    if (g.user_id !== userId || g.project !== project) continue;
    if (g.suite !== null && g.suite !== suite) continue;
    if (best === null || LEVEL_RANK[g.level] > LEVEL_RANK[best]) best = g.level;
  }
  return best;
}

/**
 * The level to publish to this caller. `null` for the classes that do not ride
 * grants at all: an admin is never filtered by one, and anonymous/static
 * callers can never hold one.
 */
function myLevel(
  viewer: SetViewer,
  project: string,
  suite: string | null,
): GrantLevel | null {
  return viewer.kind === "user"
    ? coveringLevel(viewer.id, project, suite)
    : null;
}

/** Whether this caller may see the set's runs (`visibility_predicate`). */
function visible(
  viewer: SetViewer,
  project: string,
  suite: string | null,
): boolean {
  if (viewer.kind === "full") return true;
  if (!coveringRestricted(project, suite)) return true;
  return viewer.kind === "user" && coveringLevel(viewer.id, project, suite) !== null;
}

/** Whether this caller may read/edit the set's access list (the set gate). */
export function canManageSet(
  viewer: SetViewer,
  project: string,
  suite: string | null,
): boolean {
  if (viewer.kind === "full") return true;
  if (viewer.kind !== "user") return false;
  return coveringLevel(viewer.id, project, suite) === "manage";
}

// --- aggregates ------------------------------------------------------------

interface SetRun {
  id: string;
  project: string;
  suite: string;
  created_at: number;
  pass_count: number;
  fail_count: number;
  error_count: number;
  case_count: number;
}

let SET_RUNS: SetRun[] | null = null;

/**
 * Every run that carries both a project and a suite, oldest first.
 *
 * Memoized: the corpus is generated (each run's counts come from generating its
 * cases), it never changes, and every handler below walks it — recomputing it
 * per request made the browse endpoints the slowest thing in the mock.
 */
function setRuns(): SetRun[] {
  if (SET_RUNS) return SET_RUNS;
  const out: SetRun[] = [];
  for (const r of allRunSummaries()) {
    // A run with no project can never be restricted (no row can name it) and
    // is not part of any set — the server's `project IS NOT NULL` predicate.
    if (r.project === null || r.suite === null) continue;
    out.push({
      id: r.id,
      project: r.project,
      suite: r.suite,
      created_at: parseTimestamp(r.created_at),
      pass_count: r.pass_count,
      fail_count: r.fail_count,
      error_count: r.error_count,
      case_count: r.case_count,
    });
  }
  out.sort((a, b) => a.created_at - b.created_at || a.id.localeCompare(b.id));
  SET_RUNS = out;
  return out;
}

interface Aggregate {
  run_count: number;
  last_run_at: number | null;
  pass_count: number;
  fail_count: number;
  error_count: number;
  case_count: number;
}

function aggregate(runs: SetRun[]): Aggregate {
  return runs.reduce<Aggregate>(
    (acc, r) => ({
      run_count: acc.run_count + 1,
      last_run_at: Math.max(acc.last_run_at ?? 0, r.created_at),
      pass_count: acc.pass_count + r.pass_count,
      fail_count: acc.fail_count + r.fail_count,
      error_count: acc.error_count + r.error_count,
      case_count: acc.case_count + r.case_count,
    }),
    {
      run_count: 0,
      last_run_at: null,
      pass_count: 0,
      fail_count: 0,
      error_count: 0,
      case_count: 0,
    },
  );
}

/** The server's `PASS_RATE`: a run with no cases is 0.0, never a gap. */
function runPassRate(r: SetRun): number {
  return r.case_count > 0 ? r.pass_count / r.case_count : 0;
}

/** The suite's last 20 visible pass rates, oldest first. */
function sparkline(runs: SetRun[]): number[] {
  return runs.slice(-SPARKLINE_RUNS).map(runPassRate);
}

function suiteRunsOf(
  viewer: SetViewer,
  project: string,
  suite?: string,
): Map<string, SetRun[]> {
  const bySuite = new Map<string, SetRun[]>();
  for (const r of setRuns()) {
    if (r.project !== project) continue;
    if (suite !== undefined && r.suite !== suite) continue;
    if (!visible(viewer, r.project, r.suite)) continue;
    const list = bySuite.get(r.suite) ?? [];
    list.push(r);
    bySuite.set(r.suite, list);
  }
  return bySuite;
}

// --- browse ----------------------------------------------------------------

/** `GET /sets` — every project with at least one visible run, name-ascending. */
export function listRunSets(viewer: SetViewer): SetsResponse {
  const byProject = new Map<string, SetRun[]>();
  for (const r of setRuns()) {
    if (!visible(viewer, r.project, r.suite)) continue;
    const list = byProject.get(r.project) ?? [];
    list.push(r);
    byProject.set(r.project, list);
  }

  const projects: ProjectSetView[] = [...byProject.entries()]
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([project, runs]) => {
      const agg = aggregate(runs);
      // One latest rate per visible suite, suites in name order, capped.
      const latestBySuite = new Map<string, SetRun>();
      for (const r of runs) {
        const seen = latestBySuite.get(r.suite);
        if (!seen || r.created_at >= seen.created_at) latestBySuite.set(r.suite, r);
      }
      const recent_pass_rates = [...latestBySuite.entries()]
        .sort(([a], [b]) => a.localeCompare(b))
        .map(([, r]) => runPassRate(r))
        .slice(0, PROJECT_SPARK_CAP);
      return {
        project,
        suite_count: latestBySuite.size,
        run_count: agg.run_count,
        last_run_at: agg.last_run_at,
        pass_count: agg.pass_count,
        fail_count: agg.fail_count,
        error_count: agg.error_count,
        case_count: agg.case_count,
        recent_pass_rates,
        restricted: setRestrictedExactly(project, null),
        my_level: myLevel(viewer, project, null),
      };
    });

  return { projects };
}

/** `GET /sets/{project}` — `null` (a 404) when nothing in it is visible. */
export function runSetProject(
  project: string,
  viewer: SetViewer,
): ProjectSetDetailResponse | null {
  const bySuite = suiteRunsOf(viewer, project);
  if (bySuite.size === 0) return null;

  const suites: SuiteSetView[] = [...bySuite.entries()]
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([suite, runs]) => {
      const agg = aggregate(runs);
      const series = sparkline(runs);
      return {
        suite,
        run_count: agg.run_count,
        last_run_at: agg.last_run_at,
        pass_count: agg.pass_count,
        fail_count: agg.fail_count,
        error_count: agg.error_count,
        case_count: agg.case_count,
        latest_pass_rate: series[series.length - 1] ?? null,
        sparkline: series,
        baseline_run_id: BASELINE_BY_SUITE.get(`${project}/${suite}`) ?? null,
        // Covering: a locked project locks the suites under it.
        restricted: coveringRestricted(project, suite),
        my_level: myLevel(viewer, project, suite),
      };
    });

  return {
    project,
    restricted: setRestrictedExactly(project, null),
    my_level: myLevel(viewer, project, null),
    suites,
  };
}

/** `GET /sets/{project}/suites/{suite}` — `null` (a 404) when invisible. */
export function runSetSuite(
  project: string,
  suite: string,
  viewer: SetViewer,
): SuiteSetDetailResponse | null {
  const runs = suiteRunsOf(viewer, project, suite).get(suite);
  if (!runs || runs.length === 0) return null;

  const agg = aggregate(runs);
  const series = sparkline(runs);
  return {
    project,
    suite,
    restricted: coveringRestricted(project, suite),
    my_level: myLevel(viewer, project, suite),
    run_count: agg.run_count,
    last_run_at: agg.last_run_at,
    pass_count: agg.pass_count,
    fail_count: agg.fail_count,
    error_count: agg.error_count,
    case_count: agg.case_count,
    latest_pass_rate: series[series.length - 1] ?? null,
    sparkline: series,
    baseline_run_id: BASELINE_BY_SUITE.get(`${project}/${suite}`) ?? null,
  };
}

// --- access ----------------------------------------------------------------

/** `GET /sets/.../access`. Both halves are EXACT scope — see the module head. */
export function setAccess(
  project: string,
  suite: string | null,
): SetAccessResponse {
  const rows: SetGrantView[] = grants
    .filter((g) => g.project === project && g.suite === suite)
    .sort((a, b) => a.created_at - b.created_at || a.username.localeCompare(b.username))
    .map((g) => ({
      user_id: g.user_id,
      username: g.username,
      level: g.level,
      created_at: g.created_at,
      created_by: g.created_by,
    }));
  return { restricted: setRestrictedExactly(project, suite), grants: rows };
}

/** Idempotent, like the server: re-restricting an already-locked set is a no-op. */
export function restrictSet(project: string, suite: string | null): void {
  if (!setRestrictedExactly(project, suite)) restrictions.push({ project, suite });
}

/** `false` — a 404 — when there was no restriction row to remove. */
export function unrestrictSet(project: string, suite: string | null): boolean {
  const before = restrictions.length;
  restrictions = restrictions.filter(
    (r) => !(r.project === project && r.suite === suite),
  );
  return restrictions.length !== before;
}

export function upsertGrant(
  project: string,
  suite: string | null,
  userId: string,
  username: string,
  level: GrantLevel,
  createdBy: string | null,
): void {
  const existing = grants.find(
    (g) => g.project === project && g.suite === suite && g.user_id === userId,
  );
  if (existing) {
    existing.level = level;
    return;
  }
  grants.push({
    project,
    suite,
    user_id: userId,
    username,
    level,
    created_at: Date.now(),
    created_by: createdBy,
  });
}

/** `false` — a 404 — when this user held no grant on the exact set. */
export function deleteGrant(
  project: string,
  suite: string | null,
  userId: string,
): boolean {
  const before = grants.length;
  grants = grants.filter(
    (g) => !(g.project === project && g.suite === suite && g.user_id === userId),
  );
  return grants.length !== before;
}

File: crates/domarinn-server/src/storage/sets.rs
//! Aggregates for the run-set browser (`/api/v1/sets*`).
//!
//! Separate from [`super::runs`] (which is at the per-file line ratchet) and
//! from [`super::projects`] (whose DTOs are a frozen legacy wire shape that
//! older CLIs depend on). Nothing here is shared with those endpoints on
//! purpose: the sets browser is free to change its shape, `/api/v1/projects`
//! is not.
//!
//! Every query is filtered by [`visibility_predicate`], so a project or suite
//! whose runs are all invisible disappears rather than listing with a zero
//! count — the same rule [`super::projects::list_projects`] follows.

use std::collections::HashMap;

use super::exec::{Conn, Queryable, Row, Value};
use super::projects::read_baseline;
use super::runsets::{covering_level, restricted};
use super::Storage;
use crate::dto::sets::{
    ProjectSetDetailResponse, ProjectSetView, SetsResponse, SuiteSetDetailResponse, SuiteSetView,
};
use crate::runsets::{visibility_predicate, GrantLevel, RunVisibility};

/// How many recent runs a suite's sparkline covers. Matches the legacy suite
/// series window, so the two surfaces tell the same story about "recent".
const SPARKLINE_RUNS: usize = 20;

/// How many per-suite pass rates a project's row carries. A project with more
/// suites than this shows the first 20 in name order rather than an unbounded
/// array — the row is a glance, not a report.
const PROJECT_SPARK_CAP: usize = 20;

/// The per-run aggregate columns, in the order every reader below unpacks them.
const AGGREGATES: &str = "COUNT(*), MAX(created_at), \
     CAST(COALESCE(SUM(pass_count), 0) AS BIGINT), CAST(COALESCE(SUM(fail_count), 0) AS BIGINT), \
     CAST(COALESCE(SUM(error_count), 0) AS BIGINT), CAST(COALESCE(SUM(case_count), 0) AS BIGINT), \
     CAST(COALESCE(SUM(CASE WHEN empty_count > 0 THEN empty_count ELSE 0 END), 0) AS BIGINT)";

/// A run's pass rate, as SQL. A run with no cases is 0.0 rather than NULL: the
/// sparkline is a `number[]`, and "no cases" is not a gap in the series.
const PASS_RATE: &str =
    "CASE WHEN case_count > 0 THEN CAST(pass_count AS REAL) / case_count ELSE 0.0 END";

/// `COUNT(*), MAX(created_at), SUM(pass), SUM(fail), SUM(error), SUM(cases),
/// SUM(empty)`, read from `row` starting at `offset`.
struct Aggregates {
    run_count: i64,
    last_run_at: Option<i64>,
    pass_count: i64,
    fail_count: i64,
    error_count: i64,
    case_count: i64,
    /// The `CASE` in [`AGGREGATES`] is what makes `runs.empty_count`'s
    /// tri-state safe to sum: a NULL (pre-backfill) or `-1` (undecodable blob)
    /// run adds nothing instead of poisoning the total or, worse, subtracting
    /// one from it. Zero here means "no run had anything to report", which the
    /// DTO omits — see [`crate::dto::sets`]'s header.
    empty_count: i64,
}

impl Aggregates {
    fn read(row: &Row<'_>, offset: usize) -> anyhow::Result<Aggregates> {
        Ok(Aggregates {
            run_count: row.get(offset)?,
            last_run_at: row.get(offset + 1)?,
            pass_count: row.get(offset + 2)?,
            fail_count: row.get(offset + 3)?,
            error_count: row.get(offset + 4)?,
            case_count: row.get(offset + 5)?,
            empty_count: row.get(offset + 6)?,
        })
    }

    /// The set's empty tally as the wire carries it: absent when no visible run
    /// reported one.
    fn reportable_empty_count(&self) -> Option<u64> {
        (self.empty_count > 0).then_some(self.empty_count as u64)
    }
}

impl Storage {
    /// Every project with at least one visible run, name-ascending.
    pub async fn list_run_sets(&self, vis: RunVisibility) -> anyhow::Result<SetsResponse> {
        self.runs.read(move |conn| list_run_sets(conn, &vis)).await
    }

    /// One project's suites. `None` — a 404 at the handler — when the project
    /// has no visible runs at all, so an invisible project is indistinguishable
    /// from one that never existed.
    pub async fn run_set_project(
        &self,
        project: String,
        vis: RunVisibility,
    ) -> anyhow::Result<Option<ProjectSetDetailResponse>> {
        self.runs
            .read(move |conn| run_set_project(conn, &project, &vis))
            .await
    }

    /// One suite, addressed directly. `None` under the same rule as
    /// [`Storage::run_set_project`].
    pub async fn run_set_suite(
        &self,
        project: String,
        suite: String,
        vis: RunVisibility,
    ) -> anyhow::Result<Option<SuiteSetDetailResponse>> {
        self.runs
            .read(move |conn| run_set_suite(conn, &project, &suite, &vis))
            .await
    }
}

/// A bound `TEXT` parameter, or SQL NULL when the caller left it open.
fn optional_text(value: Option<&str>) -> Value {
    match value {
        Some(text) => Value::Text(text.to_string()),
        None => Value::Null,
    }
}

/// The level to publish to this caller for `(project, suite)`.
///
/// `None` for the classes that do not ride grants at all: an admin is never
/// filtered by one, and anonymous/static-token callers can never hold one.
fn my_level(
    conn: &mut Conn<'_>,
    vis: &RunVisibility,
    project: &str,
    suite: Option<&str>,
) -> anyhow::Result<Option<GrantLevel>> {
    match vis {
        RunVisibility::User(user_id) => covering_level(conn, Some(project), suite, user_id),
        RunVisibility::Full | RunVisibility::Public => Ok(None),
    }
}

fn list_run_sets(conn: &mut Conn<'_>, vis: &RunVisibility) -> anyhow::Result<SetsResponse> {
    let mut args: Vec<Value> = Vec::new();
    let visible = visibility_predicate("runs", vis, &mut args);
    let sql = format!(
        "SELECT project, COUNT(DISTINCT suite), {AGGREGATES}
           FROM runs
          WHERE project IS NOT NULL AND {visible}
          GROUP BY project
          ORDER BY project ASC"
    );
    let rows: Vec<(String, i64, Aggregates)> = conn.query_map(&sql, &args, |row| {
        Ok((
            row.get::<String>(0)?,
            row.get::<i64>(1)?,
            Aggregates::read(row, 2)?,
        ))
    })?;

    let mut rates = latest_pass_rates(conn, vis)?;
    let mut projects = Vec::with_capacity(rows.len());
    for (project, suite_count, agg) in rows {
        let mut recent_pass_rates = rates.remove(&project).unwrap_or_default();
        recent_pass_rates.truncate(PROJECT_SPARK_CAP);
        projects.push(ProjectSetView {
            suite_count,
            run_count: agg.run_count,
            last_run_at: agg.last_run_at,
            pass_count: agg.pass_count,
            fail_count: agg.fail_count,
            error_count: agg.error_count,
            case_count: agg.case_count,
            empty_count: agg.reportable_empty_count(),
            recent_pass_rates,
            restricted: restricted(conn, Some(&project), None)?,
            my_level: my_level(conn, vis, &project, None)?,
            project,
        });
    }
    Ok(SetsResponse { projects })
}

/// Each visible `(project, suite)`'s newest run's pass rate, grouped by project
/// with the suites in name order.
///
/// One window-function pass over the whole visible corpus rather than a query
/// per project: the listing is unpaginated, so N+1 here would be N+1 over every
/// project on the instance.
fn latest_pass_rates(
    conn: &mut Conn<'_>,
    vis: &RunVisibility,
) -> anyhow::Result<HashMap<String, Vec<f64>>> {
    let mut args: Vec<Value> = Vec::new();
    let visible = visibility_predicate("runs", vis, &mut args);
    // `id` breaks ties so two runs sharing a `created_at` still pick one
    // deterministically.
    let sql = format!(
        "SELECT project, pass_rate FROM (
             SELECT project, suite, {PASS_RATE} AS pass_rate,
                    ROW_NUMBER() OVER (PARTITION BY project, suite
                                       ORDER BY created_at DESC, id DESC) AS rn
               FROM runs
              WHERE project IS NOT NULL AND suite IS NOT NULL AND {visible}
         ) AS ranked
          WHERE rn = 1
          ORDER BY project ASC, suite ASC"
    );
    let rows: Vec<(String, f64)> = conn.query_map(&sql, &args, |row| {
        Ok((row.get::<String>(0)?, row.get::<f64>(1)?))
    })?;
    let mut out: HashMap<String, Vec<f64>> = HashMap::new();
    for (project, rate) in rows {
        out.entry(project).or_default().push(rate);
    }
    Ok(out)
}

/// Whether the project has any visible run — the existence test both detail
/// endpoints 404 on.
///
/// `.query_row_opt(...)?`, never `.is_ok()`: this answer becomes a 404, so
/// swallowing the error would report "no such set" for a locked database or a
/// corrupt page, and the operator would never see the 500 that is really
/// happening.
fn project_has_visible_runs(
    conn: &mut Conn<'_>,
    project: &str,
    suite: Option<&str>,
    vis: &RunVisibility,
) -> anyhow::Result<bool> {
    let mut args: Vec<Value> = vec![project.to_string().into(), optional_text(suite)];
    let visible = visibility_predicate("runs", vis, &mut args);
    // `?2 IS NULL OR suite = ?2`: the project form ignores the suite column.
    let sql = format!(
        "SELECT 1 FROM runs
          WHERE project = ?1 AND (?2 IS NULL OR suite = ?2) AND {visible} LIMIT 1"
    );
    Ok(conn.query_row_opt(&sql, &args, |_| Ok(()))?.is_some())
}

fn run_set_project(
    conn: &mut Conn<'_>,
    project: &str,
    vis: &RunVisibility,
) -> anyhow::Result<Option<ProjectSetDetailResponse>> {
    if !project_has_visible_runs(conn, project, None, vis)? {
        return Ok(None);
    }

    let mut args: Vec<Value> = vec![project.to_string().into()];
    let visible = visibility_predicate("runs", vis, &mut args);
    let sql = format!(
        "SELECT suite, {AGGREGATES}
           FROM runs
          WHERE project = ?1 AND suite IS NOT NULL AND {visible}
          GROUP BY suite
          ORDER BY suite ASC"
    );
    let rows: Vec<(String, Aggregates)> = conn.query_map(&sql, &args, |row| {
        Ok((row.get::<String>(0)?, Aggregates::read(row, 1)?))
    })?;

    let mut sparklines = sparklines(conn, project, None, vis)?;
    let mut suites = Vec::with_capacity(rows.len());
    for (suite, agg) in rows {
        let sparkline = sparklines.remove(&suite).unwrap_or_default();
        suites.push(SuiteSetView {
            run_count: agg.run_count,
            last_run_at: agg.last_run_at,
            pass_count: agg.pass_count,
            fail_count: agg.fail_count,
            error_count: agg.error_count,
            case_count: agg.case_count,
            empty_count: agg.reportable_empty_count(),
            latest_pass_rate: sparkline.last().copied(),
            sparkline,
            baseline_run_id: read_baseline(conn, project, &suite, vis)?,
            restricted: restricted(conn, Some(project), Some(&suite))?,
            my_level: my_level(conn, vis, project, Some(&suite))?,
            suite,
        });
    }

    Ok(Some(ProjectSetDetailResponse {
        project: project.to_string(),
        restricted: restricted(conn, Some(project), None)?,
        my_level: my_level(conn, vis, project, None)?,
        suites,
    }))
}

fn run_set_suite(
    conn: &mut Conn<'_>,
    project: &str,
    suite: &str,
    vis: &RunVisibility,
) -> anyhow::Result<Option<SuiteSetDetailResponse>> {
    if !project_has_visible_runs(conn, project, Some(suite), vis)? {
        return Ok(None);
    }

    let mut args: Vec<Value> = vec![project.to_string().into(), suite.to_string().into()];
    let visible = visibility_predicate("runs", vis, &mut args);
    let sql =
        format!("SELECT {AGGREGATES} FROM runs WHERE project = ?1 AND suite = ?2 AND {visible}");
    let agg = conn.query_row(&sql, &args, |row| Aggregates::read(row, 0))?;

    let sparkline = sparklines(conn, project, Some(suite), vis)?
        .remove(suite)
        .unwrap_or_default();

    Ok(Some(SuiteSetDetailResponse {
        project: project.to_string(),
        suite: suite.to_string(),
        restricted: restricted(conn, Some(project), Some(suite))?,
        my_level: my_level(conn, vis, project, Some(suite))?,
        run_count: agg.run_count,
        last_run_at: agg.last_run_at,
        pass_count: agg.pass_count,
        fail_count: agg.fail_count,
        error_count: agg.error_count,
        case_count: agg.case_count,
        empty_count: agg.reportable_empty_count(),
        latest_pass_rate: sparkline.last().copied(),
        sparkline,
        baseline_run_id: read_baseline(conn, project, suite, vis)?,
    }))
}

/// Each suite's last [`SPARKLINE_RUNS`] visible pass rates, **oldest first**.
///
/// `suite: Some(_)` narrows to one suite; `None` covers the whole project in a
/// single pass, so the project detail does not issue a query per suite.
fn sparklines(
    conn: &mut Conn<'_>,
    project: &str,
    suite: Option<&str>,
    vis: &RunVisibility,
) -> anyhow::Result<HashMap<String, Vec<f64>>> {
    let mut args: Vec<Value> = vec![project.to_string().into(), optional_text(suite)];
    let visible = visibility_predicate("runs", vis, &mut args);
    let cap = SPARKLINE_RUNS;
    // Newest-first inside the window (that is what "the last 20" means), then
    // re-sorted oldest-first for the consumer.
    let sql = format!(
        "SELECT suite, pass_rate FROM (
             SELECT suite, created_at, id, {PASS_RATE} AS pass_rate,
                    ROW_NUMBER() OVER (PARTITION BY suite
                                       ORDER BY created_at DESC, id DESC) AS rn
               FROM runs
              WHERE project = ?1 AND suite IS NOT NULL
                AND (?2 IS NULL OR suite = ?2) AND {visible}
         ) AS ranked
          WHERE rn <= {cap}
          ORDER BY suite ASC, created_at ASC, id ASC"
    );
    let rows: Vec<(String, f64)> = conn.query_map(&sql, &args, |row| {
        Ok((row.get::<String>(0)?, row.get::<f64>(1)?))
    })?;
    let mut out: HashMap<String, Vec<f64>> = HashMap::new();
    for (suite, rate) in rows {
        out.entry(suite).or_default().push(rate);
    }
    Ok(out)
}
